app/src/main/java/org/traccar/client/StatusWidget.kt

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright 2016 - 2022 Anton Tananaev (anton@traccar.org)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.traccar.client
+
+import android.appwidget.AppWidgetManager
+import android.appwidget.AppWidgetProvider
+import android.content.*
+import android.widget.RemoteViews
+
+import androidx.preference.PreferenceManager
+
+/**
+ * Implementation of Status Widget functionality, which can display the status of the service -
+ * whether it is enabled or not.
+ */
+class StatusWidget : AppWidgetProvider() {
+
+    override fun onUpdate(context: Context, appWidgetManager: AppWidgetManager, appWidgetIds: IntArray) {
+        update(context, appWidgetManager, appWidgetIds)
+    }
+
+    override fun onEnabled(context: Context) {
+        // When the first widget is created, we want to start receiving service status broadcasts,
+        // so we can react and change the look
+        val filter = IntentFilter()
+        filter.addAction(TrackingService.ACTION_STARTED)
+        filter.addAction(TrackingService.ACTION_STOPPED)
+        // Note: we must register through the app context as a workaround
+        // for 'BroadcastReceiver components are not allowed to register to receive intents'
+        context.applicationContext.registerReceiver(Companion, filter)
+    }
+
+    override fun onDisabled(context: Context) {
+        // When the last widget is disabled (removed), we stop receiving service status broadcasts
+        context.applicationContext.unregisterReceiver(Companion)
+    }
+
+    companion object: BroadcastReceiver() {
```

Comment:
Do you mean something like this
```
@JvmStatic
val broadCastReceiver = object : BroadcastReceiver() {
   // ...
```
?

Reply:
Why does it have to be static?

Reply:
All widget's instances (objects) share same receiver.
I can add a comment to explain the need for the "companion".

Reply:
Please explain it in the comment first.

Reply:
First instance of the widget registers the receiver (to be notified of `ACTION_STARTED` and `ACTION_STOPPED`), so there is no need to create receivers for other instances of same widget. Hence one reciever's instance is enough.
Sounds reasonable?

Reply:
It doesn't really. Is it expensive to create a new instance? Sounds like premature optimization, unless I'm missing something. Also, I don't expect there to be more than one instance anyway in 99% of the cases.

Reply:
I've tried to replace `Companion` with  `val`, and... it doesn't work as expected - when single widget is removed from the screen, an exception happens inside `unregisterReceiver()` (called in `onDisabled()`), saying that the reciever isn't registered... that's strange.
I will investigate it later , when I have more time for that.
In general your change is reasonable, since there is one instance of `StatusWidget` created for all widgets.

Reply:
After some experiments I'd propose to drop the extra broadcast receiver, and register extra intents in the manifest.
Will it be acceptable to add two app-specific intents to the manifest this way
```xml
            <intent-filter>
                <action android:name="android.appwidget.action.APPWIDGET_UPDATE" />
                <action android:name="org.traccar.action.SERVICE_STARTED" />
                <action android:name="org.traccar.action.SERVICE_STOPPED" />
            </intent-filter>
```
?
I've tested this solution and it works quite good and requires less coding.

Reply:
Makes sense to me. XML configuration is always better than code.